lib/gala/payment_token.rb

```diff
@@ -37,6 +42,43 @@ def decrypt(certificate_pem, private_key_pem)
 
     class << self
 
+      def validate_signature(signature, ephemeral_public_key, data, transaction_id, application_data)
+        # Ensure that the certificates contain the correct custom OIDs
+        intermediate_cert = nil
+        leaf_cert = nil
+        p7 = OpenSSL::PKCS7.new(Base64.decode64(signature))
+        p7.certificates.each {|c|
+          c.extensions.each { |e|
+            leaf_cert = c if e.oid == LEAF_CERTIFICATE_OID
+            intermediate_cert = c if e.oid == INTERMEDIATE_CERTIFICATE_OID
+          }
+        }
+        raise InvalidSignatureError, "Signature does not contain the correct custom OIDs." unless leaf_cert && intermediate_cert
+
+        # Ensure that the root CA is the Apple Root CA - G3
+        root_cert = certificate = OpenSSL::X509::Certificate.new(APPLE_ROOT_CERT)
+
+        # Ensure that there is a valid X.509 chain of trust from the signature to the root CA
+        raise InvalidSignatureError, "Unable to verify a valid chain of trust from signature to root certificate." unless chain_of_trust_verified?(leaf_cert, intermediate_cert, root_cert)
+
+        #Ensure that the signature is a valid ECDSA signature
+        unless application_data
```

Comment:
👍
On Fri, Dec 18, 2015 at 7:41 PM Miki Rezentes <notifications@github.com>
wrote:

> In lib/gala/payment_token.rb
> <https://github.com/spreedly/gala/pull/3#discussion_r48084759>:
>
> > +        p7.certificates.each {|c|
> > +          c.extensions.each { |e|
> > +            leaf_cert = c if e.oid == LEAF_CERTIFICATE_OID
> > +            intermediate_cert = c if e.oid == INTERMEDIATE_CERTIFICATE_OID
> > +          }
> > +        }
> > +        raise InvalidSignatureError, "Signature does not contain the correct custom OIDs." unless leaf_cert && intermediate_cert
> > +
> > +        # Ensure that the root CA is the Apple Root CA - G3
> > +        root_cert = certificate = OpenSSL::X509::Certificate.new(APPLE_ROOT_CERT)
> > +
> > +        # Ensure that there is a valid X.509 chain of trust from the signature to the root CA
> > +        raise InvalidSignatureError, "Unable to verify a valid chain of trust from signature to root certificate." unless chain_of_trust_verified?(leaf_cert, intermediate_cert, root_cert)
> > +
> > +        #Ensure that the signature is a valid ECDSA signature
> > +        unless application_data
>
> If there isn't an application data, then the key is omitted. If there is
> any application data, even empty string, it could throw off the signature
> verification as I am unable to test/confirm exactly how applicationData
> needs to be added to the verification string. So, unless an empty string
> were to be falsy (which I know it doesn't in Ruby), I think it needs to
> stay as is until we have proper test data for application data.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spreedly/gala/pull/3/files#r48084759>.
>